OmniKitUI/ViewModels/InsertCannulaViewModel.swift

```diff
@@ -0,0 +1,235 @@
+            switch self {
+            case .ready:
+                return false
+            default:
+                return true
+            }
+        }
+        
+        var isProcessing: Bool {
+            switch self {
+            case .startingInsertion, .inserting:
+                return true
+            default:
+                return false
+            }
+        }
+        
+        var isFinished: Bool {
+            if case .finished = self {
+                return true
+            }
+            return false
+        }
+    }
+    
+    var error: OmnipodPumpManagerError? {
+        if case .error(let error) = self.state {
+            return error
+        }
+        return nil
+    }
+
+    @Published var state: InsertCannulaViewModelState = .ready
+    
+    var didFinish: (() -> Void)?
+    
+    var didRequestDeactivation: (() -> Void)?
+    
+    var cannulaInserter: CannulaInserter
+    
+    init(cannulaInserter: CannulaInserter) {
+        self.cannulaInserter = cannulaInserter
+    }
+    
+//    private func handleEvent(_ event: ActivationStep2Event) {
+//        switch event {
+//        case .insertingCannula:
+//            let finishTime = TimeInterval(Pod.estimatedCannulaInsertionDuration)
+//            state = .inserting(finishTime: CACurrentMediaTime() + finishTime)
+//        case .step2Completed:
+//            state = .finished
+//        default:
+//            break
+//        }
+//    }
+    
+    private func checkCannulaInsertionFinished() {
+        state = .startingInsertion
+        cannulaInserter.checkCannulaInsertionFinished() { (error) in
+            DispatchQueue.main.async {
+                if let error = error {
+                    self.state = .error(error)
+                } else {
+                    self.state = .finished
+                }
+            }
+        }
+    }
+    
+    private func insertCannula() {
+        state = .startingInsertion
+        
+        cannulaInserter.insertCannula { (result) in
+            DispatchQueue.main.async {
+                switch(result) {
+                case .success(let finishTime):
+                    self.state = .inserting(finishTime: CACurrentMediaTime() + finishTime)
+                    let delay = finishTime
+                    if delay > 0 {
+                        DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
+                            self.checkCannulaInsertionFinished() // now check if actually ready
+                        }
+                    } else {
+                        self.state = .finished
+                    }
+                case .failure(let error):
+                    self.state = .error(error)
+                }
+            }
+
+            
+//            switch status {
+//            case .error(let error):
+//                self.state = .error(error)
+//            case .event(let event):
+//                self.handleEvent(event)
+//            }
```

Comment:
[nit] still needed? (Please, consider this a general comment as there are other places where commented out code was retained).